File: Backend/dependency/application/src/test/java/com/mhp/coding/challenges/dependency/InquiryTest.java
package com.mhp.coding.challenges.dependency;

import com.mhp.coding.challenges.dependency.inquiry.Inquiry;
import com.mhp.coding.challenges.dependency.inquiry.InquiryService;
import com.mhp.coding.challenges.dependency.notifications.EmailHandler;
import com.mhp.coding.challenges.dependency.notifications.PushNotificationHandler;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.mock.mockito.SpyBean;
import org.springframework.test.context.ContextConfiguration;
import org.springframework.test.context.junit4.SpringRunner;

import static org.mockito.ArgumentMatchers.eq;
import static org.mockito.Mockito.verify;

@RunWith(SpringRunner.class)
@ContextConfiguration(classes = Application.class)
public class InquiryTest {

    @SpyBean
    private EmailHandler emailHandler;

    @SpyBean
    private PushNotificationHandler pushNotificationHandler;

    @Autowired
    private InquiryService inquiryService;

    @Test
    public void testInquiryHandlers() {
        final Inquiry inquiry = new Inquiry();
        inquiry.setUsername("TestUser");
        inquiry.setRecipient("service@example.com");
        inquiry.setText("Can I haz cheezburger?");

        inquiryService.create(inquiry);

        verify(emailHandler).sendEmail(eq(inquiry));
        verify(pushNotificationHandler).sendNotification(eq(inquiry));
    }
}

File: application/src/test/kotlin/com/mhp/coding/challenges/dependency/InquiryTest.kt
package com.mhp.coding.challenges.dependency

import com.mhp.coding.challenges.dependency.inquiry.Inquiry
import com.mhp.coding.challenges.dependency.inquiry.InquiryService
import com.mhp.coding.challenges.dependency.notifications.EmailHandler
import com.mhp.coding.challenges.dependency.notifications.PushNotificationHandler
import com.ninjasquad.springmockk.SpykBean
import io.mockk.verify
import io.mockk.verifyAll
import org.junit.jupiter.api.Test
import org.junit.jupiter.api.extension.ExtendWith
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.test.context.ContextConfiguration
import org.springframework.test.context.junit.jupiter.SpringExtension

@ExtendWith(SpringExtension::class)
@ContextConfiguration(classes = [Application::class])
class InquiryTest(
    @Autowired private val inquiryService: InquiryService,
) {

    @SpykBean
    lateinit var emailHandler: EmailHandler

    @SpykBean
    lateinit var pushNotificationHandler: PushNotificationHandler

    @Test
    fun testInquiryHandlers() {
        val inquiry = Inquiry(
            username = "TestUser",
            recipient = "service@example.com",
            text = "Can I haz cheezburger?",
        ).also { inquiryService.create(it) }

        verifyAll {
            emailHandler.sendEmail(inquiry)
            pushNotificationHandler.sendNotification(inquiry)
        }
    }
}
